src/loader/book.rs

```diff
@@ -0,0 +1,322 @@
+#![allow(missing_docs, unused_variables, unused_imports, dead_code)]
+
+use std::path::Path;
+use std::collections::VecDeque;
+use std::fs::File;
+use std::io::Read;
+
+use loader::summary::{Summary, Link, SummaryItem, SectionNumber};
+use errors::*;
+
+
+/// A dumb tree structure representing a book.
+///
+/// For the moment a book is just a collection of `BookItems`.
+#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
+pub struct Book {
+    /// The sections in this book.
+    pub sections: Vec<BookItem>,
+}
+
+impl Book {
+    /// Create an empty book.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Get a depth-first iterator over the items in the book.
+    pub fn iter(&self) -> BookItems {
+        BookItems { items: self.sections.iter().collect() }
+    }
+}
+
+/// Enum representing any type of item which can be added to a book.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum BookItem {
```

Comment:
The extra level of indirection from `BookItem` is necessary because the `SUMMARY.md` format says your book is (conceptually) a list of `Chapter`s and section separators.

The difference between `BookItem` and `SummaryItem` is because summary parsing is a separate process to loading a book. Parsing the summary should just need to take in your `SUMMARY.md` as a string and not need to touch the filesystem or load chapters into memory. That's why `Summary` is a list of `SummaryItem`s which are either section separators or `Link`s (e.g. `- [Introduction](./src/intro.md)"`). Hence the distinct `Link` entity instead of reusing `Chapter`.